Source/DependencyInversion/TenantScopedProvidersBuilder.cs

```diff
@@ -47,16 +47,20 @@ IServiceProvider CreateTenantContainer(TenantId tenant, IServiceProvider rootSer
             configure?.Invoke(tenant, services);
         }
 
+        if (services.Any(_ => _.Lifetime == ServiceLifetime.Singleton && _.ImplementationInstance == null))
+        {
+            throw new CannotRegisterSingletonDependenciesOnTenantScopedContainer();
+        }
+
         containerBuilder.Populate(services);
         containerBuilder.RegisterInstance(tenant);
         var container = containerBuilder.Build();
-
         var rootScope = container.BeginLifetimeScope(builder =>
         {
             builder.RegisterInstance(new ServiceScopeFactory(rootServiceProvider.GetRequiredService<IServiceScopeFactory>(), container)).As<IServiceScopeFactory>();
             builder.RegisterSource(new UnknownServiceOnTenantContainerRegistrationSource(rootServiceProvider));
         });
 
-        return new AutofacServiceProvider(rootScope);
+        return (new AutofacServiceProvider(rootScope), container);
```

Comment:
Related to the previous comment. It would be nice if we changed this to another implementation, maybe made a custom one so that when the service provider returned from this method disposes of the root scope when it is disposed. If it makes sense?

I might be rambling here, but it just feels like things are leaking around a little bit here...

Reply:
We can have IRootContainer or something that is disposable